File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-2.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    if head == nil {
        return nil
    }

    cur := head

    for cur != nil {
        newNode := &Node{Val: cur.Val, Next: cur.Next}
        cur.Next = newNode
        cur = newNode.Next
    }

    cur = head 
    for cur != nil {
        if cur.Random != nil {
            cur.Next.Random = cur.Random.Next
        }
        cur = cur.Next.Next
    }

    cur = head
    newHead := head.Next
    for cur != nil {
        tmp := cur.Next
        cur.Next = tmp.Next
        if tmp.Next != nil {
            tmp.Next = tmp.Next.Next
        }
        cur = cur.Next
    }
    return newHead

}

File: Data Structures & Algorithms/lru-cache/submission-2.py
class Node:
    def __init__(self, key=0, val=0):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self._move_to_head(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            self._move_to_head(node)
            node.val = value
        else:
            new_node = Node(key, value)
            self.cache[key] = new_node
            self._add_to_head(new_node)

            if len(self.cache) > self.capacity:
                lru_node = self.tail.prev
                self._remove(lru_node)

                del self.cache[lru_node.key]



    def _remove(self, node):
        prev_node = node.prev
        next_node = node.next
        prev_node.next = next_node
        next_node.prev = prev_node

    def _add_to_head(self, node):
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node
    
    def _move_to_head(self, node):
        self._remove(node)
        self._add_to_head(node)
